Ftr peak ml

sakshikukreja14 opened the pull request.

saif-el committed:
Export peak information for classification

saif-el committed:
Set-up for MOI based classification

saif-el committed:
Remove old ML logic from TableDockWidget

saif-el committed:
Remove libsvm, predictor and classifier

saif-el committed:
Basic set-up for communicating with MLB

saif-el committed:
Display probability value in peak table

saif-el committed:
Add logic for inferences from PeakML output

saif-el committed:
Add SHAP-style classification inference

saif-el committed:
Give peak group labels their own column

saif-el committed:
Allow re-classifying ML-labelled peak groups

saif-el committed:
Fix crash due to invalidated lambda contexts

saif-el committed:
[Travis] Development package for Qt Charts module

saif-el committed:
Save PeakML results in emDB

saif-el committed:
Add a new multi-selection combo box widget

saif-el committed:
Allow hiding different classes of ML prediction

saif-el committed:
Load prediction information only if available

saif-el committed:
Fix rebase error @ setLabel/setUserLabel

saif-el committed:
Fix show/hide mechanism for PeakML classes

saif-el committed:
Export classification label and probablity

saif-el committed:
Disable classification disagreement

saif-el committed:
Update PeakML icons

saif-el committed:
Enable reading in correlation information

saif-el committed:
Add correlation information to peak tables

saif-el committed:
Add table widget for correlated PeakML groups

saif-el committed:
[Fix] Assign prediction to C12 PARENT group

saif-el committed:
[UI] Hide correlation widget on table destroy

saif-el committed:
[UI] Link correlation & peak table selections

saif-el committed:
[UI] Highlight reference peak-group in correlation table

sakshikukreja14 committed:
[ftr] Added feature of undo for changing peakgroup labels

Earlier, the peakroup modified by user could not be undo to its
previous state. Added a feature where in ctrl+Z can undo the
label changed.

sakshikukreja14 committed:
[ftr] Added a feature of selecting whether to run peakMl or not

Added a checkbox in peak detection dialog, that make sure peakMl
runs after peak detection when it is checked. Otherwise, only
peaks are detected and not classified.

sakshikukreja14 committed:
[ftr] Added the login feature for polly users

Now, if the user will check the checkbox for peak curation in peak
detection dialog. The user will be redirected to the login dialog,
where in the user needs to login and has to authenticate himself
and then he will be authorized for using peakMl. If the user is
already logged in through EPI, then no login would be needed again
in order to use peakMl.

sakshikukreja14 committed:
[ftr] Modified peaktable to show  peakMl related column to authorized users

Peaktable that appear after peak detection had some columns that must
appear only when the authorized user checks for the peak curation
option. Peak table has been modified to reflect this change.

sakshikukreja14 committed:
[ftr] Added a feature of dropdown to select peakMl model Type

Added a dropdown where in the user can select the model types which
can be applied to classify the peaks in PeakMl.

sakshikukreja14 committed:
[fix] Fixed a bug in undo function

undo function misbehaved in case where peaks are marked good or bad
by the user. This bug has been fixed.

sakshikukreja14 committed:
[ftr] Removed "Explain Classification" option from context menu of peaks

when detected peaks were right clicked, the context menu appeared
the option "Explain Classification" has been removed from the menu
if the peak curation is not selected.

sakshikukreja14 committed:
[ui] Made changes to the SHAP plot that explained classification of peaks

sakshikukreja14 committed:
[fix] Fixed visualization bugs of peakgroups due to rebasing

sakshikukreja14 committed:
[fix] Fix label column in peaktable when previously saved emdb is loaded

sakshikukreja14 committed:
[fix] Fix display of peakgroups when peak curation is done after deletion an existing peakTable

sakshikukreja14 committed:
[ftr] Add correlated group information while saving session

sakshikukreja14 committed:
[ftr] Download binary and model automatically from s3 bucket

sakshikukreja14 committed:
[ftr] Download peak ml files using polly cli

sakshikukreja14 committed:
[ftr] Add peakML documentation in support dialog

sakshikukreja14 committed:
[ref] Remove qt chart modules required

sakshikukreja14 committed:
[ref] Improve force plot

sakshikukreja14 committed:
[ftr] Write cookie.json file while authenticating user to polly

Earlier, while authentiating the polly users two files(cred and
cred_refreshToken) were written. Now along with those files a
cookie.json file will be written that contains public id token
and refresh token. This is done to make calls to API by able to
send cookie along the call.

sakshikukreja14 committed:
[ftr] Classify isotopes along with parent peakgroup in peakML

Now isotopic groups are exported for classification and would
also be places in correlation table if correlated to some other
metabolites.

sakshikukreja14 committed:
[fix] Fix Group validation in status bar for good and bad peaks

sakshikukreja14 committed:
[ref] Refactor csv export options for peakgroups

sakshikukreja14 committed:
[fix] Fix bugs for peakML after rebasing to version-12(with ghost peakgroups)

sakshikukreja14 committed:
[ref] Remove 'unique id' from peakgroup

Used 'unique id' from peakgroup because earlier peakgroup id's were
not unique. Now After version 12, made groupId unique. Hence, 'unique
id' was no more required.

sakshikukreja14 committed:
[fix] Fix filtering labels and exporting peakgroups to csv

sakshikukreja14 committed:
[ftr] Add feature for 'may be good' peakgroup curation

sakshikukreja14 committed:
[ui] UI changes to range slider in peak detection dialog

sakshikukreja14 committed:
[ui] Add icon for may be good peakgroups

sakshikukreja14 committed:
[ui] UI fixes

sakshikukreja14 committed:
[fix] Fix different exports of peakgroups

sakshikukreja14 committed:
[ftr] Add legend to classification widget

sakshikukreja14 committed:
[fix] Fix crash on uploading cohort

sakshikukreja14 committed:
[ftr] Add a feature to relabel already classified groups

sakshikukreja14 committed:
[ui] Add distribution of classification percentage in the legend

sakshikukreja14 committed:
[ui] UI changes related to relabel dialog

sakshikukreja14 committed:
[ui] Add PeakML parameters in setting log

sakshikukreja14 committed:
[ref] Refactored moi download mechanism to include download from signed URL

sakshikukreja14 committed:
[ref] Changes after monday meeting

sakshikukreja14 committed:
[fix] Small fixes and ui changes

sakshikukreja14 committed:
[fix] Fix the execution of model and binary  in windows

QFile when opened in Text mode replaces '\n' to '\r\n' which
hindered the execution of peakML on windows.

sakshikukreja14 committed:
[fix] Fix closing of relabel widget on windows

sakshikukreja14 committed:
[fix] Disable removeFiles for changed API endpoint

sakshikukreja14 committed:
[fix] Fix issues occurred due to deleting commit

The commit where shap figure attributes names were normalised
has been deleted. It is decided that we would give normalised
name through python output file. Rather than mapping attributes
in El-MAVEN codebase

sakshikukreja14 committed:
[ui] Warning for no internet connection while downloading files

sakshikukreja14 committed:
[ui] Reducing the size of peak detection dialog

sakshikukreja14 committed:
[ftr] Added the list of models for organisation in peak detection dialog

sakshikukreja14 committed:
[fix] Fix crash in El-MAVEN due to cookie file

sakshikukreja14 committed:
[ref] Refactor shap figure and invert colors for signals and noise

After discussion with Kailash abut shap figure, We got to know that
left side arrows(red) are attributes that contributes to signals and
the blue for noise

saifulbkhan commented on .travis.yml:
```diff
@@ -55,7 +55,7 @@ install:
       sudo apt-get install -y -qq --allow-unauthenticated qt5-qmake qtbase5-dev qtscript5-dev qtdeclarative5-dev libqt5webkit5-dev;
       sudo apt-get install -y -qq --allow-unauthenticated --allow-downgrades libpulse-mainloop-glib0=1:8.0-0ubuntu3.10 libpulse0=1:8.0-0ubuntu3.10 libpulse-dev
-      sudo apt-get install -y -qq --allow-unauthenticated libqt5opengl5 libqt5multimedia5 libqt5multimedia5-plugins qtmultimedia5-dev;
+      sudo apt-get install -y -qq --allow-unauthenticated libqt5opengl5 libqt5multimedia5 libqt5multimedia5-plugins qtmultimedia5-dev libqt5charts5-dev;
```
Are we still using Qt Charts library? I thought we moved to a custom (QGraphicsView-based) plot for the force plot.

If we still are using it, please make sure to add this dependency in our README as well (for all three platforms).

saifulbkhan commented on src/core/libmaven/PeakGroup.cpp:
```diff
@@ -44,6 +44,7 @@ PeakGroup::PeakGroup(shared_ptr<MavenParameters> parameters,
     sampleMean=0;
 
     deletedFlag = false;
+    isHiddenFromTable = false;
```
I do not like the fact that a UI-specific flag is leaking into the core library. I know there are cases like this in the existing codebase, but we should try to minimize this with any new code.

One solution would be for each peak-table to keep a simple map of peak-group rows that are hidden and act accordingly. Feel free to ignore this if I have misunderstood the use case here.

If you choose to keep it, make it private and expose using a getter.

sakshikukreja14 replied:
Basically, we have changed the export options in CSV. One of them is exporting 'displayed groups', (which is one of the functions in csvreports.cpp, the core library) hence this flag is used there to get information if is displayed or not. 
I will make it private and use it accordingly.

saifulbkhan commented on src/core/libmaven/PeakGroup.cpp:
```diff
@@ -1069,3 +1056,93 @@ void PeakGroup::setGroupId(int groupId)
+    if (parent != nullptr && tagString == "C12 PARENT") {
+        parent->setUserLabel(label);
+        return;
+    }
```
This will never be true anymore (from v0.12.0 and above). The `tagString` of all child isotopes are always "C13-label-x", "N15-label-x", etc.

Instead, you could change the second conditional to: `parent->tagString == C12_PARENT_LABEL`, and get the same effect.

saifulbkhan commented on src/core/libmaven/PeakGroup.cpp:
```diff
@@ -1069,3 +1056,93 @@ void PeakGroup::setGroupId(int groupId)
+    for (auto& child : _childIsotopes) {
+        if (child->tagString == "C12 PARENT" && child->_userLabel != label)
+            child->setUserLabel(label);
+    }
```
This is no longer needed. The child isotope that used to have "C12 PARENT" as its `tagString` is gone.

saifulbkhan commented on src/core/libmaven/PeakGroup.cpp:
```diff
@@ -1069,3 +1056,93 @@ void PeakGroup::setGroupId(int groupId)
+    if (parent != nullptr && tagString == "C12 PARENT") {
+        parent->setPredictedLabel(label, probability);
+        return;
+    }
```
Change the second conditional as mentioned in the comment above for `PeakGroup::setUserLabel` method.

saifulbkhan commented on src/core/libmaven/PeakGroup.cpp:
```diff
@@ -1069,3 +1056,93 @@ void PeakGroup::setGroupId(int groupId)
+            && child->_predictedLabel != label
+            && child->_predictionProbability != probability)
+            child->setPredictedLabel(label, probability);
+    }
```
Again, not needed anymore. Parent peak-group is the only "C12 PARENT".

saifulbkhan commented on src/core/libmaven/PeakGroup.h:
```diff
@@ -134,19 +181,22 @@ class PeakGroup{
+        string model;
+
+        float noiseRange;
+        float mayBeRange;
```
Do these three details need to be stored per-group? Even so, I would recommend saving them in the `MavenParameters` object associated to each group. The number of things that you need to change might seem a little excessive and error prone, if you have never added a new parameter/setting to El-MAVEN before. Feel free to ask me, if you need more details around this.

saifulbkhan commented on src/core/libmaven/csvreports.cpp:
```diff
@@ -558,6 +569,72 @@ void CSVReports::_writePeakInfo(PeakGroup* group)
+                if (!child->isGhost())
+                    writePeakInfo(*child, child.get()->groupId());
+            }
+        } 
```
What if the group also has child adducts in it? They should also be exported just like child isotopes are.

saifulbkhan commented on src/core/libmaven/csvreports.cpp:
```diff
@@ -604,103 +681,103 @@ void CSVReports::writeDataForPolly(const std::string& file,
+    //     inputGroupFile.close();
+    //     savedGroupFile.close();
+    //     remove("groupReport.csv");
+    // }
```
Commented out tests until the CSV report format is final, yes? Be sure to uncomment it and make any changes required, before this PR gets merged.

sakshikukreja14 replied:
Sure, Once all the comments are resolved I will uncomment these tests.

saifulbkhan commented on src/core/libmaven/mavenparameters.cpp:
```diff
@@ -125,6 +125,9 @@ MavenParameters::MavenParameters(string settingsPath):lastUsedSettingsPath(setti
         quantileSignalBlankRatio = 0.0;
         quantilePeakWidth = 0.0;
 
+        peakMl = false;
```
We can name it something more descriptive like `usePeakMl` or `classifyUsingPeakMl`, etc. Makes it more readable when this boolean is used in an `if` conditional.

saifulbkhan commented on src/core/libmaven/mavenparameters.h:
```diff
@@ -271,6 +271,12 @@ class MavenParameters
+        //for enabling peak Ml curation
+        bool peakMl;
+        string peakMlModelType;
+        float badGroupLimit;
+        float maybeGoodGroupLimit;
```
I saw some values like this in `PeakGroup` class too. Maybe you should remove those attributes in the `PeakGroup` class and use values from each peak-group's `parameters` object.

Also why not have limits like `badGroupUpperLimit` and `goodGroupLowerLimit` and then anything in between those will be the "maybe" category? Somehow this seems like a better idea to me. But it is up to you, not necessary to change.

saifulbkhan replied:
Also I do not see any initialization for these values in "mavenparameters.cpp" (constructor). They are also not being set anywhere. Did I miss them?

sakshikukreja14 replied:
Sure will remove these from the peak-group class. Also, these values are set through peak detection dialog using slider

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -140,6 +152,24 @@ void BackgroundOpsThread::qtSignalSlot(const string& progressText,
+    emitGroups();
+
+    Q_EMIT(updateProgressBar("Done", 1, 1));
+}
```
I think I removed this function in v0.12.0. It had been there for way too long and for no reason. This is probably a rebase error. What you should do is remove this method and replace the calls to it (on lines 301 and 325) with:
```C++
    if (mainwindow->mavenParameters->peakMl) 
        classifyGroups(mavenParameters->allgroups);

    emitGroups();
```
That should do it.

saifulbkhan replied:
This is still there.

sakshikukreja14 replied:
Removed.

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -24,6 +24,11 @@
 #include "peakdetector.h"
 #include "samplertwidget.h"
 #include "EIC.h"
+#include "csvreports.h"
```
Why do we need to import CSV reporting functions here?

sakshikukreja14 replied:
Because the input file for peakML exported in this class is created using csv reports

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -356,3 +388,393 @@ void BackgroundOpsThread::updateGroups(QList<shared_ptr<PeakGroup>>& groups,
+                   + "ElMaven";
+
+    // TODO: binary name will keep changing and should not be hardcoded
+    // TODO: model should not exist anywhere on the filesystem
```
I think I added these. Are they two no longer true? If so, please remove them.

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -356,3 +388,393 @@ void BackgroundOpsThread::updateGroups(QList<shared_ptr<PeakGroup>>& groups,
+    auto mlBinary = tempDir + QDir::separator() + "moi.exe";
+#endif
+    QString mlModel;
+    // Following code is commented out because for now we have only one model in
+    // aws bucket. Once, we have more comments would be removed.  
+    // if(mainwindow->mavenParameters->peakMlModelType == "Global Model Elucidata"){
+        mlModel = tempDir + QDir::separator() + "model.pickle.dat";
+    // }
+    // else{
+    //     return;
+    // }
```
I am not aware of the complexities here, but do keep it in mind (and inform anybody who should know) that once this build is out there (in the form of a release), the behaviour cannot be changed on the fly. It can only go out as part of a new release.

sakshikukreja14 replied:
Yes, it was for testing purpose and would be removed.

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -356,3 +388,393 @@ void BackgroundOpsThread::updateGroups(QList<shared_ptr<PeakGroup>>& groups,
+                                       + "peak_ml_output.csv";
+
+    // // have to enumerated and assign each group with an ID, because multiple
+    // // groups at this point have the same ID
```
commented comments

sakshikukreja14 committed:
[ref] Accomodate the changes of output file exported through moi

Earlier Moi exported all 88 features that were used to plot
shap figure (commonly called by- inference values). Now the change
has been made in the moi file so that it exports only 10 features
that would be represented in shap figure.
This has been done for securing the IP. As all 88 features if exposed to
user would be a problem.

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -356,3 +388,393 @@ void BackgroundOpsThread::updateGroups(QList<shared_ptr<PeakGroup>>& groups,
+    int startId = 1;
+    for (auto& group : groups) {
+        group.setGroupId(startId++);
+        for (auto& child : group.childIsotopes()) {
+            child->setGroupId(startId++);
+        }
```
This for-loop should come before the line (no. 452) where you assign a group ID to the parent group. Starting v0.12.0, the `PeakGroup::setGroupId` method also updates its `metaGroupId` as well as the `metaGroupId` of its child items. But since you are setting the group ID of children after you have set it for the parent, the metaGroupId will not match. Definitely ignore this advice if `metaGroupId` has no significance here.

But do add groupID to child adducts as well. Maybe that could be important since its happening for child isotopes.

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -356,3 +388,393 @@ void BackgroundOpsThread::updateGroups(QList<shared_ptr<PeakGroup>>& groups,
+                << "--output_moi_file" << classificationOutputFile
+                << "--model_path" << mlModel
+                << "--bad_group_limit" << badGroupLimit
+                << "--maybeGood_group_limit" << maybeGoodGroupLimit;
```
Please do not mix camel-case and underscore-based naming. Choose one and stick to it. In this case please prefer underscores since that is already being used by other args.

sakshikukreja14 replied:
Made the required changes.

saifulbkhan replied:
Was also referring to `--maybeGood_group_limit` arg.

sakshikukreja14 replied:
Ohhh! Okay. To change this argument, I would need to make a change in the MOI file as the current moi takes this as an argument. 
Can we please let it be for this iteration??? As it is uploaded to Aws and folks are using it already

saifulbkhan replied:
Oh no problem. This is just a minor thing anyway.

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -356,3 +388,393 @@ void BackgroundOpsThread::updateGroups(QList<shared_ptr<PeakGroup>>& groups,
+            removeFiles();
+            return;
+        }
+        emit updateProgressBar("Classifying Peakgroups", countGroups++, groupSize);
```
Add ellipsis for indicating an ongoing process. So something like: "Classifying peak-groups…"

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -356,3 +388,393 @@ void BackgroundOpsThread::updateGroups(QList<shared_ptr<PeakGroup>>& groups,
+    for (auto& group : groups) {
+         if (mavenParameters->stop) {
+            mavenParameters->allgroups.clear();
+            terminate();
```
Why do you think the thread needs to be terminated here? What if the thread terminates before `removeFiles()` is invoked? Anyway, once you `return` from this function, the thread will stop properly, right?

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -356,3 +388,393 @@ void BackgroundOpsThread::updateGroups(QList<shared_ptr<PeakGroup>>& groups,
+                         correlations);
+    }
+    for (auto& group : groups) {
+        for (auto& child : group.childIsotopes()) {
```
Okay, one final time, most things that are happening for child isotopes also need to happen for child adducts. I know we wrote a lot of this code before v0.12.0. So, we should revisit such instances and make any relevant modifications.

I will try to keep an eye out for cases like this henceforth and mark them for your reference.

sakshikukreja14 replied:
Yup, it was on my to-do too. Adding a few commits tomorrow with this issue resolved.

sakshikukreja14 replied:
Done :)

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -356,3 +388,393 @@ void BackgroundOpsThread::updateGroups(QList<shared_ptr<PeakGroup>>& groups,
+                    + "El-MAVEN_cookie.json" ;
+    
+    ifstream readCookie(cookieFile.toStdString());
+    json cookieInput = json::parse(readCookie);
```
This might be where you need to make a change to fix that "missing cookie file" crash. Check for `QFile::exists(cookieFile)` and do something if false.

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -356,3 +388,393 @@ void BackgroundOpsThread::updateGroups(QList<shared_ptr<PeakGroup>>& groups,
+    auto dataQstring = QString::fromStdString(data);
+    QJsonObject dataObj;
+
+    QJsonDocument doc = QJsonDocument::fromJson(dataQstring.toUtf8());
+    dataObj = doc.object();  
```
So we are also using Qt JSON parsing along with the `nlohmann` library? I would have sticked to one (Qt is preferable if available and the files are not large).

sakshikukreja14 replied:
Yup, I agree with you. will make the change.

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.h:
```diff
@@ -93,8 +99,18 @@ class BackgroundOpsThread : public QThread
      */
     void restoreAlignment();
 
+    /**
+     * @brief Some functions of classification runs outside
+     * El-MAVEN and cancellation at that point can not be 
+     * handled and hence signal to disable/enable cancel button
+     * in peak detector dialog.
+     */
+    void toggleCancel(); 
```
Nice.

saifulbkhan commented on src/gui/mzroll/QHistogramSlider.h:
```diff
@@ -0,0 +1,84 @@
+    double minBound;
+    QColor colBar;
+
+};
```
I am curious what is this widget? Where are we using it?

sakshikukreja14 replied:
Sorry about this one.. I just checked it out. And it seems like some silly addition maybe because of some rebasing. I will check it further to be sure.

saifulbkhan reviewed (changes requested):
### Review part I

Added comments where I think some of the library-level code could be improved. Have yet to take a look at the front-end. Will do it in another sitting.

I did compile and try to test out basic functionality. Some comments here:
* The slider is not in the same line as the signal/noise range labels. It is also very glitchy when I drag any of the two slider controls. Clicking along the bar works well.
* The peak-detection dialog goes outside the bounds on my laptop screen (height-wise). Maybe we can have a separate PeakML tab in the detection dialog, even if it contains a handful elements. Anyhow, I feel some UI improvement can be made here.
* The application crashed for me when I set cohorts for 8 example files. I had already run one detection (with PeakML) once without setting the cohorts. Tried again, crashed on the first run - with samples cohorts set. Attached the crash log [here](https://hastebin.com/onudolagaj.yaml). Seems to be the same “missing cookie file” thing we talked about before.

Could not test further because of the crash happening every time now. Once you fix it, I will try out a few more things. Quite a bit has changed since I last worked on this branch.

sakshikukreja14 commented:
Thank you  @saifulbkhan for your reviews. Sorry, for the repeated crash. I will make the required changes soon.

saifulbkhan commented on src/gui/mzroll/forms/infodialog.ui:
```diff
@@ -25,6 +25,27 @@
+	outline: none;
+	border: 1px solid black;
+}</string>
+  </property>
```
I think we lost this style in v0.12.0. Please remove this style hunk. Probably a rebase error as well.

sakshikukreja14 committed:
[ref] Add base value and output value in classification widget

sakshikukreja14 committed:
[ref] Save output and basevalue in project DB

saifulbkhan commented on src/gui/mzroll/mainwindow.cpp:
```diff
@@ -431,6 +423,13 @@ using namespace mzUtils;
 
 	setWindowIcon(QIcon(":/images/icon.png"));
 
+    //seeting correlation table 
```
* seeting -> setting

saifulbkhan commented on src/gui/mzroll/loginform.cpp:
```diff
@@ -10,14 +10,16 @@
     _pollyelmaveninterfacedialog = pollyelmaveninterfacedialog;
     _pollyintegration = _pollyelmaveninterfacedialog->getMainWindow()->getController()->iPolly;
+
+    _showPollyApp = showPollyApp;
```
Suggestion: maybe a more descriptive name for this flag would be `showEPI`.

saifulbkhan commented on src/gui/mzroll/mainwindow.cpp:
```diff
@@ -1038,6 +1037,20 @@ void MainWindow::threadSave(const QString filename, const bool saveRawData)
+                            " connection and try again later.</b></p>");
+    htmlText += "<p>Please contact tech support at elmaven@elucidata.io if the problem persists.</p>";
+    warning->setText(htmlText);
+    warning->setIcon(QMessageBox::Icon::Critical);
```
If this is a warning, then we should be showing a "Warning" icon. Critical messages indicate that something really bad happened (e.g. major system failure or crash).

saifulbkhan commented on src/gui/mzroll/mainwindow.h:
```diff
@@ -349,14 +346,16 @@ public Q_SLOTS:
     void markGroup(shared_ptr<PeakGroup> group, char label);
     int getIonizationMode();
 	void setTotalCharge();
+    void noActiveInternet();
```
Are you sure the main-window is the best place for a slot like this? Reading this would give me the impression that this is an application-wide utility. But it is not - since it is only used for notifying about how PeakML files were not downloaded. I would suggest you either move this to peak-detection dialog or at least make  the `MainWindow::noActiveInternet` slot generic enough to be able to show any configurable message - since slots can take any number of inputs.

saifulbkhan commented on src/gui/mzroll/mzroll.pro:
```diff
@@ -79,6 +79,7 @@ CONFIG(release, debug|release) {
 INCLUDEPATH +=  /usr/include/x86_64-linux-gnu/qt5/QtXml/ /usr/include/x86_64-linux-gnu/qt5/QtSql
 
 INCLUDEPATH +=  $$top_srcdir/src/core/libmaven  \
+                $$top_srcdir/src/common         \
```
We had not added this include path earlier because all headers from the "common" namespace were included using `#include common/headername.h`. It was more indicative of the nature of library. But since we are now adding this new path, let's make sure that the include patterns are not a mix of both (i.e. - check if some places have "`common/headername.h`" and some just "`headername.h`").

saifulbkhan commented on src/gui/mzroll/numeric_treewidgetitem.cpp:
```diff
@@ -1,15 +1,37 @@
 {
     int sortCol = treeWidget()->sortColumn();
 
-    // hack to prevent sorting of child tree-widget items (such as isotopes that
-    // need to have a fixed sort order
-    if (parent() != nullptr && sortCol != 1)
-        return false;
```
Seems like another rebase issue. Please include this red diff too, it is important for sorting isotope/adduct child rows. It should be just below line no. 7 and above the newly added logic for sorting based on PeakML labels.

saifulbkhan commented on src/gui/mzroll/peakdetectiondialog.cpp:
```diff
@@ -82,8 +93,8 @@ void PeakDetectionSettings::updatePeakSettings(string key, string value)
 
-        if(QString(v.typeName()).contains("QCheckBox"))
-            v.value<QCheckBox*>()->setChecked(std::stod(value));
+/*        if(QString(v.typeName()).contains("QCheckBox"))
+            v.value<QCheckBox*>()->setChecked(std::stod(value));*/
```
Why did we comment this out? With this, checkboxes in peak-detection dialog will not be synced anymore from default settings or when loading settings from emDB files.

saifulbkhan commented on src/gui/mzroll/peakdetectiondialog.cpp:
```diff
@@ -183,6 +196,21 @@ PeakDetectionDialog::PeakDetectionDialog(MainWindow* parent) :
+        connect(peakMl, &QGroupBox::toggled,
+                [this](const bool checked)
+                {
+                    if(checked){
+                        getLoginForPeakMl();
+                    }
+                    else{
+                        peakMlSet = false;
+                        mainwindow->mavenParameters->peakMl = false;
+                        modelTypes->setEnabled(false);
+                    }
```
Let's add analytics here too? Mixpanel would be preferable over GA. Since it gives us a better idea of who was interested in PeakML. But as I mentioned before, the current label for this group-box does not really hint at some cool ML tech.

sakshikukreja14 replied:
Haven't added it till now, was facing an issue. It will be resolved asap.

sakshikukreja14 committed:
[Fix] Fix the crash when cookie file is read

saifulbkhan commented on src/gui/mzroll/peakdetectiondialog.cpp:
```diff
@@ -417,6 +506,62 @@ void PeakDetectionDialog::inputInitialValuesPeakDetectionDialog() {
+{
+    QStringList args;
+    _modelsList.clear();
+    auto _dlManager = new DownloadManager();
+    auto _pollyIntegration = new PollyIntegration(_dlManager);
+
+    auto cookieFile = QStandardPaths::writableLocation(
+                    QStandardPaths::GenericConfigLocation)
+                    + QDir::separator() 
+                    + "El-MAVEN_cookie.json" ;
+    
+    ifstream readCookie(cookieFile.toStdString());
+    if(!readCookie.is_open())
+        return;
+
+    json cookieInput = json::parse(readCookie);
+    string refreshToken = cookieInput["refreshToken"];
+    string idToken = cookieInput["idToken"];
+
+    QString cookies;
+    cookies += "refreshToken=";
+    cookies += QString::fromStdString(refreshToken);
+    cookies += ";";
+    cookies += "idToken=";
+    cookies += QString::fromStdString(idToken);
+
+    args << cookies;
+
+    auto res = _pollyIntegration->runQtProcess("listBucketObjects", args);
+    QString str(res[0].constData());
+    if (str.isEmpty())
+        return;
+    auto splitString = mzUtils::split(str.toStdString(), "\n");
+    auto data = splitString[splitString.size() - 2];
+    auto dataQstring = QString::fromStdString(data);
+    QJsonObject dataObj;
+
+    QJsonDocument doc = QJsonDocument::fromJson(dataQstring.toUtf8());
+    dataObj = doc.object();  
```
I remember seeing some very similar logic in `BackgroundOpsThread` class. Is it possible to share it? Anyhow, you should consider moving this whole work to the background thread. Reading and parsing data is a potentially blocking operation and should not be part of the main thread. Instead you could show a status string here (which I think happens post-detection when peaks are about to be classified). Or you could totally ignore all of this if you are sure that the files will never be large enough to cause an issue.

sakshikukreja14 replied:
The reason to use the same logic here is that we need to change the content of the dropdown (select models) on the basis of the Polly credentials. This is done so that the user can choose one model out of many to run PeakML.

saifulbkhan replied:
Hm, yes and the main "fetch model list" logic can still be moved to `BackgroundOpsThread` to keep all such methods in one place. But it's up to you. I do not mind repetitions in two places (I would if it was repeated thrice).

saifulbkhan commented on src/gui/mzroll/peakdetectiondialog.h:
```diff
@@ -67,6 +80,8 @@ class PeakDetectionDialog : public QDialog, public Ui_PeakDetectionDialog
                     void findPeaksClicked();
         public:
                 QString massCutoffType;
+                bool peakMlSet;
+                RangeSlider* slider; 
```
Is there a reason to keep these two public?

saifulbkhan commented on src/gui/mzroll/superSlider.h:
```diff
@@ -0,0 +1,86 @@
+#ifndef SUPERSLIDER_H
+#define SUPERSLIDER_H
+#pragma once
+
+#include <QWidget>
+#include <QPainter>
+#include <QMouseEvent>
+
+class RangeSlider : public QWidget
```
SuperSlider is a nice name. But then if the name of the class within this file is `RangeSlider`, then the filename and include guards should have the same names. Please chose one or the other. I like `RangeSlider` better - conveys exactly what it is.

saifulbkhan reviewed (changes requested):
Review part II:
* I was able to get it to not crash when I logged out of EPI and then logged back in when selecting “Peak curation” before doing peak detection. Also is “Peak curation” the right thing to call it? Should it not be something like “Use PeakML to classify peaks”? Maybe get some input from Richa and Surbhi on this.
* Even when I was able to get it to work without crashing I do not get any classification. The “Label” column was blank for all peak-groups.
* When PeakML is not able to classify anything I get “nan%” values in the label-filtering dropdown in peak tables. It should instead display “NA” or “0%”.
* Another thing that came to my mind: are we taking care of saving enough state in the emDB sessions? Such that reloading the file restores all of the data and state as it was when saved? UI state positioning and visibility excluded, of course.

Also added some more comments on the code. Please make changes wherever required.

Still need to review 15 or so file. Will do in another sitting, next week. Hopefully you will have made changes to fix some of the blocking issues.

sakshikukreja14 committed:
[ref] Classify child adducts along with the child isotopes

sakshikukreja14 committed:
[ref] Changes after review1

sakshikukreja14 committed:
[ref] Rename 'superSlider' class name to 'rangeSlider'

sakshikukreja14 committed:
[ref] Made changes after review2

sakshikukreja14 committed:
[ui] Fix peakdetection and relabel dialog

sakshikukreja14 committed:
[ref] Renaming peakML attributes in maven parameters

sakshikukreja14 committed:
[ref] Make 'noActiveInternet' of main window generic

sakshikukreja14 committed:
[ref] Store maven paramenters specific to peak curation in emdb

sakshikukreja14 committed:
[fix] Fix crash when edited peaks session is reloaded to El-MAVEN

saifulbkhan commented on .travis.yml:
```diff
@@ -87,7 +87,7 @@ script:
         # export encoded update repository URL
         export UPDATE_REPO_URL_BASE64=`echo $UPDATE_REPO_URL | base64`
-
+        
```
This is the only change in this file :P

saifulbkhan commented on src/core/libmaven/PeakGroup.cpp:
```diff
@@ -1069,3 +1059,91 @@ void PeakGroup::setGroupId(int groupId)
+    _predictionInference.clear();
+    _userLabel = label;
+
+    if (parent != nullptr && parent->tagString == C12_PARENT_LABEL) {
+        parent->setUserLabel(label);
+        return;
+    }
+
+    for (auto& child : _childIsotopes) {
+            child->setUserLabel(label);
+    }
```
As it is now, here is what I think will happen:
* If you mark an child isotope/adduct as good/bad or unmark it, then its parent will also be marked as good/bad/unmarked, right?
* When the parent's label is being set, the second loop will run and set all its child items to have the same label as the parent.

Are those things what should be happening? I think not. In isotopic labelled studies, you can very easily have a non-parent child group that has higher quality while the parent is a non-existent/bad peak-group.

In any case, you can discuss this with Richa on what should be the expected behavior. In my opinion, only mark the peak-group that is currently selected and leave its parent/siblings as they are.

sakshikukreja14 replied:
Exactly, I feel the same that child peak-group behaviour must be independent of parent/siblings. Setting the user label is outside peakML. Removing this piece of code too. So, that it gives the same functionality as it was in v0.12

saifulbkhan commented on src/core/libmaven/PeakGroup.cpp:
```diff
@@ -1069,3 +1059,91 @@ void PeakGroup::setGroupId(int groupId)
+    }
+    _predictionProbability = probability;
+
+    if (parent != nullptr && parent->tagString == C12_PARENT_LABEL) {
+        parent->setPredictedLabel(label, probability);
+        return;
+    }
+    for (auto& child : _childIsotopes) {
+        if (child->_predictedLabel != label
+            && child->_predictionProbability != probability)
+            child->setPredictedLabel(label, probability);
+    }
```
Same thing here as above. Only difference is, the model is now marking peak-groups instead of a user. The behavior should be consistent in my opinion.

sakshikukreja14 replied:
Yes, The child peak groups and child adducts are now classified separately. So, we would be requiring this piece of code. Removing this.

saifulbkhan commented on src/core/libmaven/csvreports.cpp:
```diff
@@ -558,6 +569,76 @@ void CSVReports::_writePeakInfo(PeakGroup* group)
+        }
+        for (auto& group : groups) {
+            for (auto&child : group.childIsotopes()) {
+                if (!child->isGhost())
```
Child groups are never ghosts. But okay, no harm here.

saifulbkhan commented on src/core/libmaven/mavenparameters.cpp:
```diff
@@ -309,6 +314,10 @@ void MavenParameters::copyFrom(const MavenParameters& mp)
     // vectors are meant to be used only by the global `MavenParameters` object,
     // which will not be created by a copy operation like this.
     samples = mp.samples;
+
+    peakMlModelType = mp.peakMlModelType;
+    badGroupUpperLimit = mp.badGroupUpperLimit;
+    goodGroupLowerLimit = mp.goodGroupLowerLimit;
```
Why not also copy the `MavenParameters::classifyUsingPeakMl` attribute? Makes sense to do it and you can use it to selectively show or hide PeakML attributes in "setting log" widget.

saifulbkhan commented on src/gui/mzroll/pollyelmaveninterface.cpp:
```diff
@@ -23,6 +24,8 @@ PollyElmavenInterfaceDialog::PollyElmavenInterfaceDialog(MainWindow* mw)
     qRegisterMetaType<PollyApp>("PollyApp");
     qRegisterMetaType<QMap<PollyApp, bool>>();
 
+    showPollyApps = true;
```
Is this ever set to `false`? If not, I would suggest making the default value of the second argument to the constructor of `LoginForm` to be `true`. This would make its default behavior to be what it used to and also prevent you from calling it with a flag everytime, throughout this file. A data member (`showPollyApps`) that never changes is not needed.

sakshikukreja14 replied:
Actually, this was being used while calling the log-in form and it was set to false whenever a log-in form is called through peak detection dialog. 
To get the same effect, I have removed it from the constructor and added it as a default argument in the function it was needed.

saifulbkhan commented on src/gui/mzroll/pollyelmaveninterface.cpp:
```diff
@@ -195,8 +198,9 @@ PollyElmavenInterfaceDialog::PollyElmavenInterfaceDialog(MainWindow* mw)
             this,
             &PollyElmavenInterfaceDialog::_changeMode);
+    connect(this, SIGNAL(loginResponse()), _mainwindow->peakDetectionDialog, SLOT(loginSuccessful()));
+    connect(this, SIGNAL(loginUnsuccessful()), _mainwindow->peakDetectionDialog, SLOT(unsuccessfulLogin()));
```
Please use the new signal-slot syntax for all newly written code. It is type-safe and fails fast.

saifulbkhan commented on src/gui/mzroll/pollyelmaveninterface.cpp:
```diff
@@ -419,6 +423,27 @@ void PollyElmavenInterfaceDialog::_enableExistingProjectUi()
 
+void PollyElmavenInterfaceDialog::loginForPeakMl()
+{
+    try {
+        _callLoginForm(false);
+    } catch (...) {
+        QMessageBox msgBox(this);
+        msgBox.setWindowModality(Qt::NonModal);
+        msgBox.setWindowTitle("Error in loading login form");
+        msgBox.exec();
+    } 
```
Why did we need a try catch here? The login form has it's own error reporting capabilities, right?

sakshikukreja14 replied:
Yes, but those include if the user is not authorized or email Id not given. This try-catch block reports if for some reason the login form fails to appear (the probability of which is very low).

saifulbkhan commented on src/gui/mzroll/pollyelmaveninterface.cpp:
```diff
@@ -1087,9 +1113,9 @@ QStringList PollyElmavenInterfaceDialog::_prepareFilesToUpload(QDir qdir,
+    peakTable->exportJsonToPolly(_writeableTempDir, jsonFilename, true);
 
     if (_selectedApp == PollyApp::PollyPhi) {
-        peakTable->exportJsonToPolly(_writeableTempDir, jsonFilename, true);
```
Why did this move outside the `if` block?

saifulbkhan commented on src/gui/mzroll/pollyelmaveninterface.cpp:
```diff
@@ -419,6 +423,27 @@ void PollyElmavenInterfaceDialog::_enableExistingProjectUi()
+void PollyElmavenInterfaceDialog::loginFormClosed()
+{
+    emit loginUnsuccessful();
+}
```
Hmm... I see these being called as a slot by `LoginForm`. And then these signals are themselves used to trigger slots in `PeakDetectionDialog`. Why not directly connect the `LoginForm` signals to `PeakDetectionDialog` slots? Eliminate this seeming unnecessary layer?

saifulbkhan commented on src/gui/mzroll/rangeSlider.h:
```diff
@@ -0,0 +1,86 @@
+#ifndef RANGESLIDER_H
+#define RANGESLIDER_H
+#pragma once
```
I know you borrowed this code from an external source, but let's stick to include guards for consistency and drop the pragma.

saifulbkhan commented on src/gui/mzroll/tabledockwidget.cpp:
```diff
@@ -211,8 +327,15 @@ shared_ptr<PeakGroup> TableDockWidget::groupForItem(QTreeWidgetItem *item)
+  if (hasClassifiedGroups)
+    index = 1;
+  else
+    index = 0;
```
Set `index` with a default value of `0` and then if `hasClassifiedGroups` is true, then set it to `1`. Ensures default init value, conveys the same intent and lesser code.

saifulbkhan commented on src/gui/mzroll/tabledockwidget.cpp:
```diff
@@ -296,89 +517,140 @@ void TableDockWidget::updateItem(QTreeWidgetItem *item, bool updateChildren) {
-      item->setText(15, QString::number(group->changeFoldRatio, 'f', 3));
-      item->setText(16, QString::number(group->changePValue, 'f', 6));
+      item->setText(index++, QString::number(group->changeFoldRatio, 'f', 3));
+      item->setText(index++, QString::number(group->changePValue, 'f', 6));
```
Really good stuff! Using an incrementing index to set items instead of absolute values. Why did I never thing of this...

saifulbkhan commented on src/gui/mzroll/tabledockwidget.cpp:
```diff
@@ -678,51 +959,171 @@ void TableDockWidget::showAllGroups() {
+
+    // Setting curation parameters;
+    if (hasClassifiedGroups) {
+      group->isClassified = true;
+      for (auto childIsotope : group->childIsotopes()) {
+        childIsotope->isClassified = true;
+      }
```
Child adducts?

sakshikukreja14 replied:
Added.

saifulbkhan commented on src/gui/mzroll/tabledockwidget.cpp:
```diff
@@ -855,12 +1256,22 @@ void TableDockWidget::exportGroupsToSpreadsheet() {
                         includeSetNamesLines,
                         _mainwindow->mavenParameters);
-  QList<shared_ptr<PeakGroup>> selectedGroups = getSelectedGroups();
+  QList<shared_ptr<PeakGroup>> selectedGroups;
+  
+  if (static_cast<int>(peakTableSelection) == 10)
+    selectedGroups = getDisplayedGroups();
+  else
+    selectedGroups = getSelectedGroups();
```
Does that `10` mean something special here? Please use a constant or at least add a comment.

saifulbkhan commented on src/gui/mzroll/tabledockwidget.cpp:
```diff
@@ -1471,6 +2102,30 @@ void TableDockWidget::unmarkGroup() {
   _mainwindow->autoSaveSignal(currentGroups);
 }
 
+// bool TableDockWidget::checkLabeledGroups() {
+
+//   int totalCount = 0;
+//   int goodCount = 0;
+//   int badCount = 0;
+
+//   if (_mainwindow->peaksMarked >= allgroups.size()) {
+//     for (int i = 0; i < allgroups.size(); i++) {
+//       char groupLabel = allgroups[i].userLabel();
+//       if (groupLabel == 'g') {
+//         goodCount++;
+//       } else if (groupLabel == 'b') {
+//         badCount++;
+//       }
+//       totalCount++;
+//     }
+
+//     if (totalCount == goodCount + badCount)
+//       return true;
+//   }
+
+//   return false;
+// }
```
No longer used? Why not remove it entirely?

saifulbkhan commented on src/gui/mzroll/tabledockwidget.cpp:
```diff
@@ -1799,7 +2482,17 @@ void TableDockWidget::contextMenuEvent(QContextMenuEvent *event) {
+  {
+      QAction *z8 = menu.addAction("Explain classification");
+      classificationWidget = new ClassificationWidget(this);
+      connect(z8,
+              SIGNAL(triggered()),
+              classificationWidget,
+          SLOT(showClassification()));
```
Use type-safe syntax for signal-slot.

saifulbkhan commented on src/gui/mzroll/tabledockwidget.cpp:
```diff
@@ -1865,6 +2560,84 @@ void TableDockWidget::writeMascotGeneric(QString filename) {
+      && changedMaybeGoodLimit == maybeGoodGroupLimit) {
+    QMessageBox::warning(this,
+                          "No Change!",
+                          "The curation parameters you entered are same as" 
+                          " in classified table.\nYou may change the range"
+                          " and try again.\n");
```
"The curation parameters you entered are the same as they were in the classified peak-table."

"You may change the range and try again."

saifulbkhan commented on src/gui/mzroll/tabledockwidget.cpp:
```diff
@@ -1865,6 +2560,84 @@ void TableDockWidget::writeMascotGeneric(QString filename) {
+  int totalSteps = groups.size();
+  int steps = 1;
+  for (auto &group : groups) {
+    Q_EMIT(updateRelabelStatusBar("Relabeling", steps, totalSteps));
```
"Relabeling…"

saifulbkhan commented on src/gui/mzroll/tabledockwidget.cpp:
```diff
@@ -2827,6 +3658,7 @@ void ScatterplotTableDockWidget::setupPeakTable() {
     colNames << "Max quality";
     colNames << "MS2 score";
     colNames << "#MS2 events";
+    colNames << "Probability"; // TODO: add this column conditionally
```
Don't you think we should name this "Classification probability" instead? Ignore if this decision was made after discussion already.

sakshikukreja14 replied:
Sure..Will discuss it with Richa and change it accordingly.

saifulbkhan commented on src/pollyCLI/pollyintegration.cpp:
```diff
@@ -35,7 +36,7 @@ PollyIntegration::PollyIntegration(DownloadManager* dlManager):
     #endif
 
-    indexFileURL = "https://raw.githubusercontent.com/ElucidataInc/polly-cli/master/prod/index.js";
+    indexFileURL = "https://raw.githubusercontent.com/sakshikukreja14/polly-cli/ftr_moi_api/prod/index.js";
```
I am going to keep this comment up so that the branch does not get merged without this change.

saifulbkhan reviewed (changes requested):
Review part III - leaving some more comments here. Will do user-based testing now based on the feature document that you shared earlier. Might do this today itself.

saifulbkhan commented:
@sakshikukreja14 A few comments from previous reviews have not been resolved because I do not see the requested changes. Do take a look.

sakshikukreja14 commented:
@saifulbkhan sure will resolve the comments soon. Sorry, for missing out on some previously.

sakshikukreja14 committed:
[ref] Made changes after review 3

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -267,8 +279,14 @@ void BackgroundOpsThread::computePeaks()
+
     emitGroups();
 
+    Q_EMIT(updateProgressBar("Done", 1, 1));
+
```
This should not be here. The emit will be followed by another emit in the next line, effectively making it useless.

saifulbkhan commented on src/gui/mzroll/backgroundopsthread.cpp:
```diff
@@ -290,8 +308,14 @@ void BackgroundOpsThread::findFeatures()
+
     emitGroups();
 
+    Q_EMIT(updateProgressBar("Done", 1, 1));
+
```
Same as above.

saifulbkhan commented on src/gui/mzroll/pollyelmaveninterface.cpp:
```diff
@@ -13,6 +13,7 @@
 #include "pollywaitdialog.h"
 #include "projectdockwidget.h"
 #include "tabledockwidget.h"
+#include "peakdetectiondialog.h"
```
This is no longer needed here now that we have moved the slot connection to login-form, right?

saifulbkhan commented on src/gui/mzroll/pollyelmaveninterface.h:
```diff
@@ -191,6 +204,10 @@ public Q_SLOTS:
      */
     void uploadFinished(bool success);
 
+    void loginResponse();
+
+    void loginUnsuccessful();
+
```
These two should be removed now.

saifulbkhan commented:
@sakshikukreja14 Adding another usability review for PeakML:
* Peak detection dialog comes up with “Polly-PeakML” tab selected by default. Should be “Detection method”.
* The “Preparing inputs for classification…” step is very slow for large peak-count. Is it the write-to-CSV part? I am on an SSD. Normal hard-disks can frustrate user. Eventually I force quit the application - not sure if this was just stuck or slow.
* Tried the same dataset again in a fresh session and it worked very fast this time.
* Sorting by label column does not work. Is that intentional?
* Cycling through the “correlated peak-groups” worked well. But the currently selected peak-group in the peak-table was always the first entry in the correlation table. So if I cycle between masses 99, 100, 101 and 102 and the current mass selected is 101, then the order in the correlation table is 101, 99, 100, 102. Would be nice to have them always sorted by mass. The cycling itself is according to mass (which is the right behavior). So there is some inconsistency.
* I edited one of the maybe peak-groups to include some of the missed peak areas and then marked it as “good” manually. Then pressed Cmd+z to see if undo is working. It worked the first time but was really slow (UI froze for about 5 seconds). Then I marked it as good again and then again tried undoing the marking this time it froze for 5 seconds again but did not undo my manually set “good” mark. Tried undo (pressed Cmd+z) one more time and then it worked. This was a table with less-than 200 peak-groups.
* The force plot can do a little better when it comes to drawing labels for contributing attributes. In the example below, the markers are overlapping with the text (on the blue side) but there’s ample space to the right that can be used to position them without any overlaps.
<img width="1012" alt="Screenshot 2020-12-23 at 8 25 19 AM" src="https://user-images.githubusercontent.com/10397162/102955432-e2f32e00-44fb-11eb-9b1c-c12e5bf78378.png">

* I am able to select multiple peak-groups in the table and then explain their classification (but the force plot only comes up for the last selected group). We should disable the “Explain classification” menu item during multi-selection.
* After relabeling the peak-table once or twice using new ranges, the context-menu (that comes up on right-click) for peak-groups no longer came up again for that table. It worked fine on any newly created table. Was not able to reproduce this bug again. Might be a one off case.

Everything else seems to be working as it should. The QA team might have some more feedback since they will probably play around with it more than I could.

sakshikukreja14 committed:
[fix] Made changes after usability review

sakshikukreja14 committed:
[ref] Preparation of implementing license for peakML

sakshikukreja14 committed:
[fix] Divide 'Preparing inputs for peakML' in 3 small steps

This is done because moi sometime takes time to run depending on
the no. of peaks, and user might think the system is hanged.
Hence, showing progress at different steps

sakshikukreja14 committed:
[fix] Fix Polly-phi working in peakML version

Polly-phi didn't work because of the extra classified label in peak
tables. This has been corrected by removing extra columns in case of
polly export

sakshikukreja14 committed:
[ftr] Add event tracker to peakML checkbox in peakdetectin dialog

sakshikukreja14 committed:
[fix] update the peakML range in setting logs after relabeling

sakshikukreja14 committed:
[ref] Change the predicted_label column name and add extra column

Change the predicted_label column name to peakML_label_id and added
an extra column that would export labels in string format.
This is done because the upt add in downstream needed this extra
column

sakshikukreja14 committed:
[ref] Refactored the logic for downloading model files including license check

sakshikukreja14 committed:
[ftr] Add version check for binary file 'moi'

sakshikukreja14 committed:
[fix] Fix json check on Api response

sakshikukreja14 committed:
[fix] Fix losing information of infenece plot while loading emdb

sakshikukreja14 committed:
[ref] Change the name of the columns in emdb schema for peakML

sakshikukreja14 committed:
[ftr] Export peakML related attributes in csv both in group and peak report

sakshikukreja14 committed:
[ref] Change variable names of baseValue and outputValue in peakgroup class

sakshikukreja14 committed:
[fix] Fix unauthorized error message for fetching models

sakshikukreja14 committed:
[ftr] Add an input dialog for user input to save EIC in raw data emDB

sakshikukreja14 committed:
[ftr] Integrate the logic of setting rtMin and rtMax for storing EIC in raw data emDB

sakshikukreja14 committed:
[ftr] Add an option on EPI to export emDB with raw data

sakshikukreja14 committed:
[ftr] Cohort file made necessity for accessing peakML

Added a warning for the user if he tries to access peakML without
uploading the cohort file.

sakshikukreja14 committed:
[ref] Change export raw data options in input dialog

sakshikukreja14 committed:
[ftr] Allow user to choose uploading sliced or complete EIC using EPI

sakshikukreja14 committed:
[ref] Make 'complete-rangeEIC' option on EPI by default checked

sakshikukreja14 committed:
[ref] Change the integration of test polly to prod polly

sakshikukreja14 committed:
[Fix] Fix crash on peak detection using SRM database

sakshikukreja14 committed:
[fix] Fix issues after rebasing

sakshikukreja14 committed:
[ref] Add Parallelization in curation

sakshikukreja14 committed:
[fix] Label legend on classified peak table doesnt maintain its filtered state

sakshikukreja14 committed:
[fix] Marking peak group good or bad pushes the group down

When sorted on label, marking group good or bad pushes it down on
the sorted positon.
It is fixed by enabling and disabling sorting.
